pkg/operation/botanist/component/etcd/etcd.go

```diff
@@ -265,20 +279,83 @@ func (e *etcd) Deploy(ctx context.Context) error {
 		return err
 	}
 
-	if _, err := controllerutils.GetAndCreateOrMergePatch(ctx, e.client, networkPolicy, func() error {
-		networkPolicy.Annotations = map[string]string{
+	// add peer certs if shoot has HA control plane
+	var (
+		etcdPeerCASecret *corev1.Secret
+		peerServerSecret *corev1.Secret
+	)
+	if _, ok := e.annotations[v1beta1constants.ShootAlphaControlPlaneHighAvailability]; ok && gardenletfeatures.FeatureGate.Enabled(features.HAControlPlanes) {
+		etcdPeerCASecret, found = e.secretsManager.Get(v1beta1constants.SecretNameCAETCDPeer)
+		if !found {
+			return fmt.Errorf("secret %q not found", v1beta1constants.SecretNameCAETCDPeer)
+		}
+
+		peerServerSecret, err = e.secretsManager.Generate(ctx, &secretutils.CertificateSecretConfig{
+			Name:                        secretNamePrefixPeerServer + e.role,
+			CommonName:                  "etcd-server",
+			DNSNames:                    e.serviceDNSNames(),
+			CertType:                    secretutils.ServerClientCert,
```

Comment:
Why is this `ServerClientCert` again? Shouldn't it be `ServerCert` only?

Reply:
No. The peer certificates are used for two-way communication - both server and client authentication is done by the peers on the peer network, so we require this certificate to be of type `ServerClientCert`.

Reply:
Can you explain how the rotation of the peer CA will work then (without introducing disruption/downtime) (see also [GEP-18](https://github.com/gardener/gardener/blob/master/docs/proposals/18-shoot-CA-rotation.md#rotation-sequence-of-other-cas))?

Reply:
@rfranzke Thanks for the link. I think the following approach should work for peer certs in the CA rotation story:
```
- t0
  - etcd trusts client certificates signed by ETCD_CA0 and uses a server certificate signed by ETCD_CA0
  - in case of a multi-node etcd setup, etcd also trusts and uses peer certificates (server-client certificates) signed by ETCD_CA0
  - kube-apiserver and backup-restore use a client certificate signed by ETCD_CA0 and trust ETCD_CA0
- t1:
  - Generate ETCD_CA1
  - Update etcd:
    - Trust CLIENT CERTS signed by both ETCD_CA0 and ETCD_CA1
    - Renew PEER (SERVERCLIENT) CERT (sign it with ETCD_CA1)
  - Update kube-apiserver and backup-restore:
    - Adapt CA bundle to trust both ETCD_CA0 and ETCD_CA1
    - Renew CLIENT CERTS (sign them with ETCD_CA1)
- t2:
  - Update etcd:
    - Trust only CLIENT CERTS signed by ETCD_CA1
    - Renew SERVER CERT (sign it with ETCD_CA1)
  - Update kube-apiserver and backup-restore so that their CA bundle contains only ETCD_CA1
```

Unfortunately this will now require three restarts of etcd.
- any change to the etcd members will be applied in a rolling fashion.
- If we renew the peer certs at `t1`, that means the following sequence of events will occur:
  - `t1.1`
    - `Member0` is updated and now trusts certs signed by both CA0 and CA1. It also now uses new peer cert signed by CA1.
    - `Member1` and `member2` still only trust certs signed by CA0, so they can no longer authenticate `member0`.
    - `Member0` is still able to authenticate `member1` and `member2`, but is not part of the etcd cluster anymore.
    - If `member0` was previously leader, then leadership will transfer to either `member1` or `member2`.
    - The etcd cluster state still maintains quorum as `member1` and `member2` are able to communicate with each other, unless a node/zone failure occurs at this point.
  - `t1.2`
    - `Member1` is updated and now trusts certs signed by both CA0 and CA1. It also now uses new peer cert signed by CA1.
    - `Member0` and `member1` are now able to freely communicate with each other, hence they form a quorum amongst themselves.
    - `Member2` is no longer part of the etcd cluster, as it cannot be authenticated by `member0` and `member1`.
    - If `member2` was previously leader, leadership will now be transferred to either `member0` or `member1`.
  - `t1.3`
    - `Member2` is updated and now trusts certs signed by both CA0 and CA1. It also now uses new peer cert signed by CA1.
    - Now all members can freely communicate with each other as all of them now trust certs signed by both CA0 and CA1, and use peer certs signed by CA1.
    - The etcd cluster is now fully healthy, and is using new peer certs that are valid on all members.

Of course, all this is based on theory as of now. We definitely need to test this hypothesis out to make sure it really works this way and quorum is maintained throughout this process.

Also, the above approach works in a happy-path case - as long as there are no zone failures during this period, especially for a multi-zonal etcd deployment. If a zone fails during `t1`, then the etcd cluster will lose quorum and etcd-druid will have to kick in to revive it. If we want to avoid such cases and make the peer CA rotation completely bulletproof, we will need to introduce another step between `t1` and `t2` (let's call it `t1.5` for now), where the renewal of peer certificate happens during this `t1.5` step. This would mean we introduce another restart of the etcd, effectively having a total of 3 etcd restarts for the CA rotation for etcd.

What are your thoughts on this?
cc @unmarshall @abdasgupta

Reply:
Yes, I don't see a way around three etcd restarts in total during CA rotation:
- 2 restarts in phase one to not break the existing cluster (first restart only populates new CAs to all members, second restart renews the certificates)
- 1 restart in phase two (dropping the old CAs)